feat: Add high level api for timestamp conversion

OussamaSaoudi opened the pull request:
## What changes are proposed in this pull request?

NOTE: Only look at the latest PR

This PR introduces high-level timestamp conversion APIs for Delta tables, allowing conversion between timestamps and table versions.
 - Add `latest_version_as_of(snapshot, engine, timestamp)` - returns the latest version at or before a
  timestamp
  - Add `first_version_after(snapshot, engine, timestamp)` - returns the first version at or after a
  timestamp
  - Add `timestamp_range_to_versions(snapshot, engine, start, end)` - converts a timestamp range to a
  version range

  These APIs enable timestamp-based time travel queries and are the public interface for the
  history_manager module.



**Stack:**
- PR #899
- PR #900 

## How was this change tested?
Unit tests verify timestamp to version conversion functionality with both file modification timestamps and in-commit timestamps. Tests cover edge cases including exact matches, out-of-range timestamps, and transitions between timestamp types.

zachschuermann commented on kernel/src/history_manager/mod.rs:
```diff
@@ -44,14 +49,13 @@ type Timestamp = i64;
+pub struct LogHistoryManager {
     log_segment: LogSegment,
     snapshot: Arc<Snapshot>,
+    commit_to_timestamp_cache: RefCell<HashMap<Url, Timestamp>>,
```
does a cache belong in kernel?

OussamaSaoudi replied:
Yeah I wondered about that as well. I decided to go with caching because the cache is c*O(size of log segment). And the constant c is rly small.

zachschuermann commented on kernel/src/table.rs:
```diff
@@ -76,6 +77,46 @@ impl Table {
+    /// - `limit`: Optional maximum number of versions to track. When specified, the earliest
+    ///   queryable version will be `snapshot.version() - limit`. This allows trading
+    ///   memory usage for historical reach.
+    pub fn history_manager_from_snapshot(
```
considering we have two constructors here already and we have optional info in each of them, would it make more sense to just have a builder?
```rust
let hm = HistoryManagerBuilder::new(engine)
    .withSnapshot(snapshot)
    .withLimit(1_000)
    .build()
```

aside: I wonder if it's more idiomatic to have engine in `new` or `build`..?

and should we include end version?

OussamaSaoudi replied:
Now all functions are static, so we avoid this :)

OussamaSaoudi committed:
feat: add low-level timestamp-to-version conversion

Add core infrastructure for timestamp-based time travel:
- timestamp_to_version: converts timestamp to version using binary search
- get_timestamp_search_bounds: determines search strategy based on ICT enablement
- TimestampSearchBounds: enum for search scenarios
- Simplified LogHistoryError types (collapsed TimestampOutOfRangeError into enum)
- Made Bound public for use in error types

Includes debug tracing for search bounds and ICT reads.

Co-authored-by: Isaac

OussamaSaoudi committed:
feat: add debug assertions for search bounds validation

Add debug_assert! statements to verify lo and hi indices are within valid
bounds before array access in timestamp_to_version.

Co-authored-by: Isaac

OussamaSaoudi committed:
feat: add monotonization for file modification timestamp searches

File modification timestamps are not guaranteed to be monotonically
increasing due to clock skew or filesystem limitations. This change:

- Adds linear_search_file_mod_timestamps() which monotonizes timestamps
  during search (if ts <= prev_ts, adjust to prev_ts + 1)
- Uses binary search only for ICT-based searches (ICT is guaranteed
  monotonic by the Delta protocol)
- Uses linear scan with monotonization for file modification searches

This matches the behavior of delta-kernel-java's DeltaHistoryManager.

Co-authored-by: Isaac

DrakeLin commented on kernel/src/history_manager/mod.rs:
```diff
@@ -1 +1,734 @@
+    // Get the index of the ICT enablement version in the log segment. If the version is not
+    // present, binary_search returns the insertion point (first commit at or after).
+    let commit_count = log_segment.listed.ascending_commit_files.len();
+    let ict_enablement_idx = log_segment
```
can we check whether the first commit's version > ict_enablement_version and return `ICTSearchStartingFrom(0)` directly instead of going through binary search for tables with cleaned up commits?

sanujbasu replied:
should we defer the search to be after the match arm?

sanujbasu replied:
also can the binary search for the ict index be replaced by simple arithmetic since the versions are contiguous? We can perhaps do that if we have invariants validating there are no holes and no compaction files?

OussamaSaoudi replied:
> We can perhaps do that if we have invariants validating there are no holes and no compaction files?

that should already  part of the `LogSegment::for_timestamp_conversion`.


> can we check whether the first commit's version > ict_enablement_version and return ICTSearchStartingFrom(0) 

makes sense.

DrakeLin commented on kernel/src/history_manager/mod.rs:
```diff
@@ -1 +1,734 @@
+            }
+
+            let lo_version = commit_range[0].version;
+            let hi_version = commit_range[commit_range.len() - 1].version;
```
```suggestion
            let hi_version = commits.last().unwrap().version;
```

DrakeLin commented on kernel/src/history_manager/mod.rs:
```diff
@@ -1 +1,734 @@
+    }
+
+    let lo_version = commits[0].version;
+    let hi_version = commits[commits.len() - 1].version;
```
```suggestion
    let hi_version = commits.last().unwrap().version;
```

OussamaSaoudi replied:
clippy doesn't allow unwrap

DrakeLin commented on kernel/src/history_manager/mod.rs:
```diff
@@ -1 +1,734 @@
+/// # Errors
+/// Returns [`LogHistoryError::TimestampOutOfRange`] if the timestamp is out of range. This can
+/// happen in the following cases based on the bound:
+/// - `Bound::GreatestLower`: There is no commit whose timestamp is lower than the given
```
```suggestion
/// - `Bound::GreatestLower`: There is no commit whose timestamp is less than or equal to the given
```

DrakeLin commented on kernel/src/history_manager/mod.rs:
```diff
@@ -1 +1,734 @@
+/// happen in the following cases based on the bound:
+/// - `Bound::GreatestLower`: There is no commit whose timestamp is lower than the given
+///   `timestamp`.
+/// - `Bound::LeastUpper`: There is no commit whose timestamp is greater than the given `timestamp`.
```
```suggestion
/// - `Bound::LeastUpper`: There is no commit whose timestamp is greater than or equal to the given `timestamp`.
```

DrakeLin commented on kernel/src/history_manager/mod.rs:
```diff
@@ -1 +1,734 @@
+        }
+    }
+
+    // Table: v0=50, v1=150, v2=250 (file mod), v3=300, v4=400 (ICT at v3)
```
i think some of the test comments are off here

DrakeLin reviewed (approved):
looks good, have a couple nits and tests to consider adding:
 
1. No test with ict_enablement_version: None (file-mod-only table). The FileModificationSearch path is never exercised. 
2. No test with ict_enablement_version: Some(0) (ICT from creation). The ICTSearchStartingFrom(0) from Enabled { enablement: None } is never exercised.
3. No test with non-monotonic file-mod timestamps. The monotonization logic (raw_ts.max(prev_monotonic_ts + 1)) is never triggered.
4. No test querying a timestamp between ICT commits (e.g. 350, between v3=300 and v4=400). The ICT binary search for mid-range values is untested.

OussamaSaoudi committed:
docs: add monotonization trade-offs comment

Document the trade-offs of file modification timestamp monotonization
in linear_search_file_mod_timestamps, including edge cases near
retention boundaries and the relationship with ICT as the long-term
solution.

Co-authored-by: Isaac

OussamaSaoudi committed:
chore: revert unrelated formatting change

Co-authored-by: Isaac

OussamaSaoudi committed:
refactor: make Bound internal by storing reason string in error

Change TimestampOutOfRange error to store a reason string instead of
the Bound enum, keeping Bound as an internal implementation detail.

Co-authored-by: Isaac

sanujbasu commented on kernel/src/schema/diff.rs:
```diff
@@ -2625,57 +2548,36 @@ mod tests {
-                                                "renamed_inner_key", // Renamed!
-                                                DataType::INTEGER,
-                                                false,
-                                                3,
```
why are these removed? What are the diffs?

sanujbasu commented on kernel/src/error.rs:
```diff
@@ -225,6 +225,10 @@ pub enum Error {
+
+    /// Error during log history operations (timestamp queries, version lookups)
+    #[error(transparent)]
+    LogHistory(#[from] Box<crate::history_manager::error::LogHistoryError>),
```
well something area has custom logging :)

sanujbasu commented on kernel/src/history_manager/search.rs:
```diff
@@ -13,12 +13,24 @@ use std::fmt::Debug;
 /// * [`Bound::GreatestLower`] - Finds the largest index `i` such that `values[i] <= key`. This
 ///   represents the last element less than or equal to the search key.
 #[allow(unused)]
```
This `#[allow(unused)]` is now stale

sanujbasu commented on kernel/src/history_manager/search.rs:
```diff
@@ -13,12 +13,24 @@ use std::fmt::Debug;
+
 /// Represents the errors that can occur when performing binary search using
 /// [`binary_search_by_key_with_bounds`].
 #[allow(unused)]
```
Stale `#[allow(unused)]. Drop the attribute.

sanujbasu commented on kernel/src/history_manager/search.rs:
now used?

OussamaSaoudi replied:
removed all the allow(unused)

sanujbasu commented on kernel/src/history_manager/mod.rs:
```diff
@@ -1 +1,1344 @@
+    // The ICT enablement index must be within the log segment. If it equals commit_count,
+    // the ICT enablement version is beyond all commits in the segment.
+    require!(
+        ict_enablement_idx < commit_count,
+        LogHistoryError::internal(
```
1/        
```
 ict_enablement_idx < commit_count,
        LogHistoryError::internal(
            "ICT enablement version beyond log segment",
            DeltaError::generic(format!(
                "ICT enablement version {} not in log segment (index {} >= count {})",
                ict_enablement_version, ict_enablement_idx, commit_count
            ))
```
could be wrapped into `LogHistoryError::ICTEnablementBeyondLogSegment { enablement_version, commit_count }`

2/ Could log maintenance operations break this invariant so that it becomes a case we need to support? Do we need to do a linear file modification time based search instead of erroring?

OussamaSaoudi replied:
> could be wrapped into LogHistoryError::ICTEnablementBeyondLogSegment { enablement_version, commit_count }


I'd agrue this is more of an invariant violation. this case says ICT enablemnet is in the future. that's a corrupt table and we should not try to fallback. Let me update the error msg.

sanujbasu commented on kernel/src/history_manager/mod.rs:
```diff
@@ -1 +1,1344 @@
+/// Given a timestamp, this function determines the commit range that timestamp conversion
+/// should search. A timestamp search may be conducted over one of two version ranges:
+///     1) A range of commits whose timestamp is the file modification timestamp
+///     2) A range of commits whose timestamp is an in-commit timestamp.
```
since this is a critical function lets be more  descriptive about the bounds returned in the method description and the errors thrown?

sanujbasu commented on kernel/src/history_manager/mod.rs:
```diff
@@ -1 +1,1344 @@
+    let result = match timestamp.cmp(&ict_timestamp) {
+        Ordering::Equal => TimestampSearchBounds::ExactMatch(ict_enablement_version),
+        Ordering::Less => TimestampSearchBounds::FileModificationSearchUntil {
+            index: ict_enablement_idx,
```
super critical function :) . Lets document match arm behavior for future reviewers/ readers (human or not)

```
  // Per PROTOCOL.md §In-Commit Timestamps
  //   Rule 1: ts >= enablementTs => consider only versions >= enablementVersion
  //   Rule 2: ts <  enablementTs => consider only versions <  enablementVersion
  // Equal: enablementTs IS the ICT of enablementVersion ), so this
  //   is the exact answer for both bounds.
  // Less: restrict search to the pre-ICT region via FileModificationSearchUntil.
  //   Spec guarantees enablementTs > prev_mod_time, so when the
  //   pre-ICT linear scan fails LeastUpper, enablementVersion is the true LUB.
  // Greater: restrict search to ICT commits only.

```

sanujbasu commented on kernel/src/history_manager/mod.rs:
```diff
@@ -1 +1,1344 @@
+    for commit in commits {
+        let raw_ts = commit.location.last_modified;
+        // Monotonize: ensure each timestamp is strictly greater than the previous
+        let monotonic_ts = raw_ts.max(prev_monotonic_ts + 1);
```
defensively

`raw_ts.max(prev_monotonic_ts.saturating_add(1));`

sanujbasu commented on kernel/src/history_manager/mod.rs:
```diff
@@ -1 +1,1344 @@
+            Bound::LeastUpper => {}
+        }
+
+        prev_monotonic_ts = monotonic_ts;
```
can we break and set the result from one place? Every match arm seems to have custom behavior

sanujbasu replied:
```suggestion
  /// Yields (version, monotonized_timestamp) pairs, scanning commits in order.
  /// The monotonization rule matches DeltaHistoryManager.java:533-554:
  /// strictly increasing by at least 1ms over visible history.
  fn monotonized_timestamps<'a>(
      commits: &'a [ParsedLogPath],
  ) -> impl Iterator<Item = (Version, Timestamp)> + 'a {
      let mut prev = i64::MIN;
      commits.iter().map(move |c| {
          let mono = c.location.last_modified.max(prev.saturating_add(1));
          if mono != c.location.last_modified {
              warn!(version = c.version, raw = c.location.last_modified, mono, "...");
          }
          prev = mono;
          (c.version, mono)
      })
  }

  fn linear_search_file_mod_timestamps(
      commits: &[ParsedLogPath],
      timestamp: Timestamp,
      bound: Bound,
  ) -> Result<Version, LogHistoryError> {
      let mut iter = monotonized_timestamps(commits);
      match bound {
          Bound::GreatestLower => {
              iter.take_while(|&(_, ts)| ts <= timestamp)
                  .last()
                  .map(|(v, _)| v)
                  .ok_or(LogHistoryError::TimestampOutOfRange {
                      timestamp, reason: bound.out_of_range_reason(),
                  })
          }
          Bound::LeastUpper => {
              iter.find(|&(_, ts)| ts >= timestamp)
                  .map(|(v, _)| v)
                  .ok_or(LogHistoryError::TimestampOutOfRange {
                      timestamp, reason: bound.out_of_range_reason(),
                  })
          }
      }
  }
```
would something like the above be a cleaner impl of the function?

OussamaSaoudi replied:
would something like the above be a cleaner impl of the function?

take while + last and all the monadic operations feel a bit less intuitive to me

sanujbasu commented on kernel/src/history_manager/mod.rs:
```diff
@@ -1 +1,1344 @@
+        } => {
+            let commits = &log_segment.listed.ascending_commit_files[..index];
+            linear_search_file_mod_timestamps(commits, timestamp, bound).or_else(|e| {
+                // For LeastUpper, if no version found, the ICT enablement version is the answer
```
Pattern-match the specific case? 

```
  .or_else(|e| match (e, bound) {
      (LogHistoryError::TimestampOutOfRange { .. }, Bound::LeastUpper) => Ok(ict_enablement_version),
      (e, _) => Err(e),
  })
```

sanujbasu commented on kernel/src/history_manager/mod.rs:
```diff
@@ -1 +1,1344 @@
+    timestamp: Timestamp,
+    bound: Bound,
+) -> Result<Version, LogHistoryError> {
+    // Get log segment for timestamp conversion
```
could short circuit the log segment read here like the java kernel and delta spark?

```
  let snap_ts = snapshot.timestamp(engine)?;  // cached on the snapshot
  match (timestamp.cmp(&snap_ts), bound) {
      (Ordering::Equal, _) => return Ok(snapshot.version()),
      (Ordering::Greater, Bound::GreatestLower) => return Ok(snapshot.version()),
      (Ordering::Greater, Bound::LeastUpper) => {
          return Err(LogHistoryError::TimestampOutOfRange {
              timestamp, reason: bound.out_of_range_reason(),
          });
      }
      _ => {}
  }

```

sanujbasu commented on kernel/src/history_manager/mod.rs:
```diff
@@ -1 +1,1344 @@
+    // Get log segment for timestamp conversion
+    let log_segment = LogSegment::for_timestamp_conversion(
+        engine.storage_handler().as_ref(),
+        snapshot.log_segment().log_root.clone(),
```
LogSegment::for_timestamp_conversion drops StagedCommit. Do we need to handle this?

sanujbasu replied:
If true, should we error if not supported?

OussamaSaoudi replied:
Added a check and todo #2443

sanujbasu commented on kernel/src/history_manager/mod.rs:
```diff
@@ -1 +1,1344 @@
+        engine.storage_handler().as_ref(),
+        snapshot.log_segment().log_root.clone(),
+        snapshot.version(),
+        None,
```
We rebuild a log segment via for_timestamp_conversion instead of reusing snapshot.log_segment(). Is it because for_timestamp_conversion returns contiguous commits only while snapshot.log_segment() may include a checkpoint prefix with gaps.
 
If so, should we add a code comment?

OussamaSaoudi replied:
is this sufficient?
```
    // Build a dedicated log segment for timestamp conversion. We cannot reuse
    // snapshot.log_segment() because it may include a checkpoint prefix with gaps in
    // the commit sequence. for_timestamp_conversion returns only contiguous commits,
    // which is required for correct timestamp-to-version mapping.
```

sanujbasu replied:
yes. Thank you

sanujbasu commented on kernel/src/history_manager/mod.rs:
```diff
@@ -1 +1,1344 @@
+
+        // ICT search: use binary search (ICT timestamps are guaranteed monotonic by protocol)
+        TimestampSearchBounds::ICTSearchStartingFrom(lo) => {
+            let commit_range = &log_segment.listed.ascending_commit_files[lo..];
```
can we split into a helper so that this logic reads more like prose?

```
  match search_bounds {
      ExactMatch(v) => Ok(v),
      ICTSearchStartingFrom(lo) => binary_search_ict_timestamps(&commits[lo..], timestamp, bound, engine),
      FileModificationSearch => linear_search_file_mod_timestamps(&commits, timestamp, bound),
      FileModificationSearchUntil { index, ict_enablement_version } => {
          linear_search_file_mod_timestamps(&commits[..index], timestamp, bound)
              .or_else(|e| match (e, bound) {
                  (LogHistoryError::TimestampOutOfRange { .. }, Bound::LeastUpper) => Ok(ict_enablement_version),
                  (e, _) => Err(e),
              })
      }
  }

```

OussamaSaoudi replied:
nice suggestion. updated

sanujbasu commented on kernel/src/history_manager/mod.rs:
````diff
@@ -1 +1,1344 @@
+/// ```
+#[allow(unused)]
+#[tracing::instrument(skip(snapshot, engine), fields(latest_version = snapshot.version(), table_root = %snapshot.table_root()))]
+pub(crate) fn latest_version_as_of(
````
`version_at_or_before` ?

OussamaSaoudi replied:
iirc this is the naming in spark.

sanujbasu commented on kernel/src/history_manager/mod.rs:
````diff
@@ -1 +1,1344 @@
+/// ```
+#[allow(unused)]
+#[tracing::instrument(skip(snapshot, engine), fields(latest_version = snapshot.version(), table_root = %snapshot.table_root()))]
+pub(crate) fn first_version_after(
````
version_at_or_after ?

sanujbasu commented on kernel/src/history_manager/mod.rs:
````diff
@@ -1 +1,1344 @@
+/// let (start_version, end_version) =
+///     timestamp_range_to_versions(&snapshot, &engine, start_timestamp, Some(end_timestamp))?;
+/// ```
+#[allow(unused)]
````
truly unused? Maybe run an AI audit and remove?

sanujbasu commented on kernel/src/history_manager/mod.rs:
````diff
@@ -1 +1,1344 @@
+///     timestamp_range_to_versions(&snapshot, &engine, start_timestamp, Some(end_timestamp))?;
+/// ```
+#[allow(unused)]
+#[tracing::instrument(skip(snapshot, engine), fields(latest_version = snapshot.version(), table_root = %snapshot.table_root()))]
````
latest_version every is actually the pinned snapshot version right?

sanujbasu commented on kernel/src/history_manager/mod.rs:
```diff
@@ -1 +1,1344 @@
+    }
+
+    // Convert the start timestamp to version
+    let start_version = first_version_after(snapshot, engine, start_timestamp)?;
```
since both first_version_after and latest_version_as_of perform listing won't some thing like

```
  let log_segment = LogSegment::for_timestamp_conversion(...)?;
  let start_version = timestamp_to_version_in_segment(snapshot, engine, &log_segment, start_timestamp, Bound::LeastUpper)?;
  let end_version = end_timestamp
      .map(|end| timestamp_to_version_in_segment(snapshot, engine, &log_segment, end, Bound::GreatestLower))
      .transpose()?;

```

where `timestamp_to_version_in_segment` is a helper help cut listing time in half?

sanujbasu replied:
Are there perf histograms capturing this behavior?

OussamaSaoudi replied:
> where timestamp_to_version_in_segment is a helper help cut listing time in half?


yes. This is rly unoptimized and has a long way to go. I think we can get a first cut in and optimize. There's a boatload left on the table.

OussamaSaoudi replied:
> Are there perf histograms capturing this behavior?

No, future work to add metrics. baby steps :)

sanujbasu replied:
Thanks for filing the TODO #2449

sanujbasu commented on kernel/src/history_manager/mod.rs:
```diff
@@ -1 +1,1344 @@
+                    start_timestamp,
+                    end_timestamp: end,
+                    between_version: end_version,
+                })
```
2 ways of contructing errors? into() above and from() here? Reconcile?

sanujbasu reviewed (changes requested):
Thanks for pushing on this. Left some comments.

OussamaSaoudi committed:
feat: add high-level timestamp-to-version conversion APIs

Add public APIs for timestamp-based time travel queries:
- `latest_version_as_of`: Get the latest version at or before a timestamp
- `first_version_after`: Get the first version at or after a timestamp
- `timestamp_range_to_versions`: Convert a timestamp range to version range

Includes monotonization for file modification timestamps to handle clock
skew, and comprehensive boundary tests covering standard tables, file-mod-only
tables, and ICT-from-creation tables.

Co-authored-by: Isaac

OussamaSaoudi committed:
refactor: make LogHistoryError::Internal source optional

Allow creating internal errors with just a context message (no underlying
source error). This simplifies error creation for cases where there's no
wrapped error.

Co-authored-by: Isaac

OussamaSaoudi committed:
feat: make high-level timestamp APIs public with tracing

- Change latest_version_as_of, first_version_after, and
  timestamp_range_to_versions from pub(crate) to pub
- Remove #[allow(unused)] annotations since functions are now public
- Add ret to tracing instrumentation for result logging
- Clean up test comments for consistency across test functions

Co-authored-by: Isaac

OussamaSaoudi committed:
fix: address kernel review findings for timestamp APIs

- Make Timestamp type public (was pub(crate)) so users see the doc
  comment explaining milliseconds since epoch
- Fix module-level docs to remove misleading "construction" reference
- Fix first_version_after example comment to say "at or after"
- Simplify doc comments for latest_version_as_of and first_version_after
- Add test for ICT fast path when pre-ICT commits are cleaned up

Co-authored-by: Isaac

OussamaSaoudi committed:
fix: make history_manager module unconditionally public

The high-level timestamp-to-version APIs are intended to be public,
not gated behind the internal-api feature. This change:
- Makes the history_manager module unconditionally pub in lib.rs
- Makes the error submodule unconditionally pub so users can handle
  LogHistoryError

Co-authored-by: Isaac

OussamaSaoudi committed:
perf: short-circuit timestamp lookup using snapshot timestamp

Compare against snapshot's timestamp before rebuilding the log segment.
This mirrors Delta Spark and Java Kernel behavior:
- If timestamp equals snapshot timestamp: return snapshot version
- If timestamp > snapshot timestamp with GreatestLower: return snapshot version
- If timestamp > snapshot timestamp with LeastUpper: error (no version exists)

This avoids the expensive log segment construction when the answer is
already available from the snapshot.

Co-authored-by: Isaac

OussamaSaoudi committed:
refactor: extract binary_search_ict_timestamps helper

Simplify the match in timestamp_to_version by extracting ICT binary
search logic into a dedicated helper function. This makes the main
match read more like prose, matching the style of the existing
linear_search_file_mod_timestamps helper.

Co-authored-by: Isaac

OussamaSaoudi committed:
fix: make history_manager module pub(crate) with allow(unused)

The low-level timestamp APIs are internal implementation details.
- Change history_manager module from internal-api conditional to pub(crate)
- Add #![allow(unused)] to all module files since they're unused until
  the high-level PR exposes them publicly

Co-authored-by: Isaac

OussamaSaoudi committed:
docs: add protocol-based documentation for ICT search bounds match arms

Document each match arm in get_timestamp_search_bounds with protocol
reasoning for future reviewers (human or otherwise).

Co-authored-by: Isaac

OussamaSaoudi committed:
refactor: change non-monotonic timestamp log from warn to trace

Non-monotonic file modification timestamps are expected in some
environments and don't indicate a problem - they're handled correctly
by monotonization. Downgrade to trace level to avoid noisy logs.

Co-authored-by: Isaac

OussamaSaoudi committed:
chore: add TODO for log segment reuse optimization

Add note about potential optimization to reuse the log segment between
start and end timestamp conversions in timestamp_range_to_versions.

Co-authored-by: Isaac

sanujbasu reviewed (approved):
thanks for making the changes and filing the TODOs
